### ISSUE TITLE
Improvements to URL list management for web_connectivity

### ISSUE THREAD

**Issue by hellais** [ooni/probe-android#105]
We would like to improve how URL lists are managed inside of web_connectivity. The main things we would like to support are:

1. Allowing users to choose which categories they are going to be testing when running ooniprobe on mobile
2. Testing also the country specific lists together with the global list

Before outlining my proposes solution I would like to take a step back and thing of this from a user perspective.

As a ooniprobe user when I choose the categories to be tested and click on the web connectivity "Run" button what I am most interested in is:

a. If any site of the categories I have picked there is blocked on my network
b. Contribute more measurements of that type to the ooniprobe dataset

Given the limitations on mobile we don't actually scan the full list but in most cases (with the default timeout of 90 seconds) will likely be testing only about a dozen sites in total.

URL lists are by nature something that is dynamic so in coming up with a solution for this we should take this into account and it's probably not ideal to have to make a new app release in order to update the testing lists (this is why ooniprobe on desktop, for example, has a dynamic resources update system).

In light of a., maybe we don't actually want to give users all URLs to test, but maybe we want to give them a set that includes some known blocked websites (to give them gratification when they find them to be blocked by running the app) and also some URLs that have not been tested, to improve the ooni dataset (b.).

Given these requirements and reasoning, this is starting do sound a lot like probe orchestration. So what if the running of web connectivity was simply a special orchestration request for accepting a task that is going to be run immediately?

The workflow would look something like the following:

* User configures some settings and chooses which categories of sites they want to test
* Client sends request to orchestration system saying "I am from country XX, I want to test these categories and my test runtime is YY, give me some URLs"
* Orchestration system compiles a list of URLs that includes only the categories chosen by the user, long enough to take at least YY to run, but not too long to not waste bandwidth and returns it to the probe
* The probe takes the URLs and runs the test

Thoughts?

**Comment by lorenzoPrimi** [comment]
Good reasoning, I like it. Updating test lists would have been such an issue, expecially in terms of bandwidth.
But it's not the "orchestration" as for me it means "push notifications" and an user can have them disabled. It's more a backend feature, right?

Furthermore in the app there will be the options to manually enter an URL or retest the same set you just tested so test lists are not needed locally for this options.

**Comment by hellais** [comment]
> But it's not the "orchestration" as for me it means "push notifications" and an user can have them disabled. It's more a backend feature, right?

Correct it would not work via push notifications, but it would mean that before running the web_connectivity test the app would contact the orchestration backend (`events.proteus.ooni.io`) and ask for a task to run.

I call it orchestration, because functionally it's very similar to what a probe does when it receives a push notification to do some task, except it would not have:

a. Signing of messages (since it's user initiated I don't consider it that risky to have things not be signed)
b. All the workflow of `accept`, `reject`, `done` that orchestration has, as it's a one-shot operation.

> Furthermore in the app there will be the options to manually enter an URL or retest the same set you just tested so test lists are not needed locally for this options.

Yes, this is however a separate issue I think.

**Comment by bassosimone** [comment]
@hellais I think the idea is brilliant. Ideally, I'd like it to be implemented in such a way that we can compose the existing `WebConnectivity` functionality and orchestration neatly. That would probably mean initiating a specific orchestration request and then reacting to the response similarly to what we would do when we receive a notification for testing a specific URL. (If that sounds to you equal to what you said, good, because it means I've understood what you mean and we're on the same page, otherwise, let's discuss this further in terms of how you would like it to land inside of measurement-kit.)

**Comment by hellais** [comment]
>  I'd like it to be implemented in such a way that we can compose the existing WebConnectivity functionality and orchestration neatly. That would probably mean initiating a specific orchestration request and then reacting to the response similarly to what we would do when we receive a notification for testing a specific set of URLs

Yes that sounds about right. I think the only differences to when you receive a notification are going to be:

a. The request is not signed (we want this to be the case, because signing requires the keys to be hardcoded into the app and we can't enforce, like we can with notifications, that clients of only a specific version receive certain messages)
b. It doesn't necessarily need to follow the `accept`, `reject`, `done` workflow. Though this is just for convenience, not due to technical limitations.

**Comment by lorenzoPrimi** [comment]
@hellais if you want you can close it, please mention the relative commit for orchestra
In Android is this https://github.com/ooni/probe-android/commit/7fad32c7b3bb85fc71cfe20c5d84b3e51365d0fa

**Comment by hellais** [comment]
Let's leave this open until we switch to the final canonical URL `orchestrate.ooni.io`.

This is going to happen as soon as: https://github.com/ooni/orchestra/pull/50 lands and we have deployed it to production.

It will at this point likely happen early next week.

**Comment by hellais** [comment]
This is now deployed on orchestrate.ooni.io.

In light of https://github.com/ooni/orchestra/issues/51 I have made a minor change to the URL, be sure to check your code to all the `test-list` path before `urls`, like so:

https://orchestrate.ooni.io/api/v1/test-list/urls?country_code=IT

**Comment by hellais** [comment]
I left a couple of comment in https://github.com/ooni/probe-android/commit/7fad32c7b3bb85fc71cfe20c5d84b3e51365d0fa of what changes are needed. Close this ticket and link the commit when it's updated in probe-android.

**Comment by lorenzoPrimi** [comment]
https://github.com/ooni/probe-android/commit/ccf21442b24cc084b16f415cb0bb87e6c3f9c4d9
